lib/services/vote_animation_service.dart

```diff
@@ -0,0 +1,135 @@
+        : (initialState == VoteAnimation.voted ||
+                initialState == VoteAnimation.voteFocused ||
+                initialState == VoteAnimation.voteUnfocused)
+            ? (type == VoteButtonType.down) ? downVotedColor : votedColor
+            : unfocusedColor;
+  }
+
+  Future voteItem(Vote vote) async {
+    if (vote == lastVote) {
+      vote = vote == Vote.favorite ? Vote.up : Vote.none;
+    } else if (vote == Vote.up && lastVote == Vote.favorite) vote = Vote.none;
+    _setStates(vote);
+    lastVote = vote;
+  }
+
+  void addButtonStateListener(
+      VoteButtonType type, VoteStateChangedHandler onStateChange) {
+    if (buttonStates[type] == null)
+      buttonStates[type] = ValueNotifier<VoteAnimation>(VoteAnimation.initial)
+        ..addListener(() => onStateChange(buttonStates[type].value));
+    _setState(type, lastVote == null ? initialVote : lastVote);
+  }
+
+  void _setStates(Vote vote) =>
+      buttonStates.keys.forEach((type) => _setState(type, vote));
+
+  void _setState(VoteButtonType type, Vote vote) => buttonStates[type].value =
+      _getState(type.toVote(), buttonStates[type].value, vote);
+
+  VoteAnimation _getState(Vote button, VoteAnimation currentState, Vote vote) {
+    if ([
+      VoteAnimation.voted,
+      VoteAnimation.voteFocused,
+      VoteAnimation.voteUnfocused,
+    ].contains(currentState))
+      return _reduceState(
+        vote: vote,
+        button: button,
+        whenFocused: VoteAnimation.clearFocused,
+        whenVoted: VoteAnimation.voted,
+        whenElse: VoteAnimation.clearUnfocused,
+      );
+
+    if ([
+      VoteAnimation.focused,
+      VoteAnimation.clearFocused,
+    ].contains(currentState))
+      return _reduceState(
+        vote: vote,
+        button: button,
+        whenFocused: VoteAnimation.focused,
+        whenVoted: VoteAnimation.voteFocused,
+        whenElse: VoteAnimation.unfocused,
+      );
+
+    if ([
+      VoteAnimation.unfocused,
+      VoteAnimation.clearUnfocused,
+    ].contains(currentState))
+      return _reduceState(
+        vote: vote,
+        button: button,
+        whenFocused: VoteAnimation.focused,
+        whenVoted: VoteAnimation.voteUnfocused,
+        whenElse: VoteAnimation.unfocused,
+      );
+
+    assert(currentState == VoteAnimation.initial);
+    return _reduceState(
+      vote: vote,
+      button: button,
+      whenFocused: VoteAnimation.focused,
+      whenVoted: VoteAnimation.voted,
+      whenElse: VoteAnimation.unfocused,
+    );
+  }
+
+  VoteAnimation _reduceState({
+    Vote vote,
+    Vote button,
+    VoteAnimation whenFocused,
+    VoteAnimation whenVoted,
+    VoteAnimation whenElse,
+  }) =>
+      _shouldFocus(vote)
+          ? whenFocused
+          : _shouldVote(vote, button) ? whenVoted : whenElse;
+
+  bool _shouldFocus(Vote vote) => vote == Vote.none;
+
+  bool _shouldVote(Vote vote, Vote button) =>
+      vote == button || vote == Vote.favorite && button == Vote.up;
```

Comment:
Use braces to make it more readable. It should be either
`(vote == button || vote == Vote.favorite) && button == Vote.up;`
or
`vote == button || (vote == Vote.favorite && button == Vote.up);`
depending on the intent

Reply:
fixed with 27e8d8d intended as written && over ||